Remove teleblock timer in safe zones

Nightfirecat opened the pull request:
Fixes runelite/runelite#2384

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/timers/TimersPlugin.java:
```diff
@@ -123,6 +175,35 @@ public void onVarbitChange(VarbitChanged event)
+	@Subscribe
+	public void onWidgetHiddenChanged(WidgetHiddenChanged event)
+	{
+		if (worldType == PVP)
```
This check is present because entering the wilderness on a normal world will trigger this condition as most of the pvp-related widgets load un-hidden and are thereafter hidden as needed. (that is, the safe zone widget is loaded un-hidden then immediately hidden before the frame renders)

With that in mind, world hopping between PVP worlds may need to be checked for as this same condition may be met there.

Nightfirecat replied:
After testing, my suspicion here was correct--hopping between PVP worlds in a non-safe zone with this current code will register a safe zone trigger because of this phenomenon. Unfortunately, the `GameState` is `LOGGED_IN` by the time it happens, so this can't be gated behind a check for that. Does anyone have an idea of how to tackle this issue? To my knowledge, I can't add an async check for this, otherwise I'd just add some kind of short delay (say, 100ms or less) to check the widget state at that point.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/timers/PvpWorldType.java:
```diff
@@ -0,0 +1,51 @@
+
+import net.runelite.api.WorldType;
+
+enum PvpWorldType
```
I would somehow reduce the need to basically have this structure what is really similar to one in XP tracker.

Nightfirecat replied:
Mentioned in discord, but I can't think of a good way to do this without completely rewriting WorldType, which seems like overengineering it.

deathbeam replied:
```java
static list <WorldType> DIFF = { DMM, SDMM, TOURNAMENT } 
static list <WorldType> PVP = { PVP }.join(DIFF)
boolean WorldType#isDifferentAccountWorld(WorldType other)
{
return 
  // hopping to different acc world
  (DIFF.contains(other) && !DIFF.contains(this)) || 
  // hopping from different acc world
  (!DIFF.contains(other) && DIFF.contains(this))
}

boolean WorldType#isPvp()
{
  return PVP.contains(this)
}
```

something like this probably should be enough

Nightfirecat replied:
It was a little trickier than that (hopping between `DIFF` types also required clearing tbs), but I went with that approach.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/timers/TimersPlugin.java:
```diff
@@ -505,18 +553,64 @@ public void onGameTick(GameTick event)
+				if (worldType == DMM || lastWorldType == DMM
+					|| worldType == SDMM || lastWorldType == SDMM)
+				{
+					log.debug("Hopped to or from a DMM world, clearing Teleblock timer.");
```
This also probably needs to account for TOURNAMENT worlds

Nightfirecat replied:
Tournament worlds are now included in the "different account type" worlds.

Nightfirecat commented:
Updated to remove the world hop checks as they're effectively handled upon the logout check introduced in #5090. The only two conditions the updated code now checks for are:

1. Exiting wilderness on a non-PVP world
2. Entering safe zone on a PVP world

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/timers/TimersPlugin.java:
```diff
@@ -720,4 +779,34 @@ private void removeGameTimer(GameTimer timer)
+
+	private static boolean isPvpWorld(final EnumSet<WorldType> worldTypes)
+	{
+		return listContainsAnyFromSet(PVP_WORLD_TYPES, worldTypes);
```
return worldTypes.stream().anyMatch(w -> PVP_WORLD_TYPES.contains(w));

Nightfirecat replied:
Huh, didn't know you could stream sets. Will do.

deathbeam commented:
Needs rebase

Nightfirecat commented:
Rebased

Nightfirecat commented:
Derp, fixed that mistake.

Nightfirecat commented:
Added better doc comment to `IN_WILDERNESS` varbit, fixed wilderness widget IDs, and removed some spammy false positives on the "entered PVP safe zone" condition. This should be completed and ready to go.

deathbeam commented on runelite-api/src/main/java/net/runelite/api/WorldType.java:
```diff
@@ -75,6 +76,13 @@
 		this.mask = mask;
 	}
 
+	public static EnumSet<WorldType> PVP_WORLD_TYPES = EnumSet.of(
```
Maybe this should be private? or maybe each world could have flag if its PVP and you would build this enumset based on that

Nightfirecat replied:
Wasn't sure about exactly how to go about this--just knew I wanted to move it into WorldType since I'll actually need this for another PR. I'd be fine with making it private or making it an enum flag, though I slightly prefer just making it private.

deathbeam replied:
Alright I guess we can go with that

Nightfirecat replied:
Done

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/timers/TimersPlugin.java:
```diff
@@ -506,18 +541,35 @@ public void onGameTick(GameTick event)
-		if (!currentWorldPoint.equals(lastPoint))
+		widgetHiddenChangedOnPvpWorld = false;
+
+		Widget widget = client.getWidget(PVP_WORLD_SAFE_ZONE.getGroupId(), PVP_WORLD_SAFE_ZONE.getChildId());
```
client.getWidget(PVP_WORLD_SAFE_ZONE) ?

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/timers/TimersPlugin.java:
```diff
@@ -527,10 +579,12 @@ public void onGameStateChanged(GameStateChanged gameStateChanged)
+			removeTbTimers();
+		}
+
+		if (gameStateChanged.getGameState() == GameState.LOGGED_IN)
```
else if (maybe you should change this entire method to do switch (gameStateChanged.getGameState()) instead, it will be cleaner

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/timers/TimersPlugin.java:
```diff
@@ -506,18 +541,35 @@ public void onGameTick(GameTick event)
+		widgetHiddenChangedOnPvpWorld = false;
+
+		Widget widget = client.getWidget(PVP_WORLD_SAFE_ZONE.getGroupId(), PVP_WORLD_SAFE_ZONE.getChildId());
+		if (widget == null)
```
just merge this check with the next one probably

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/timers/TimersPlugin.java:
```diff
@@ -146,6 +158,32 @@ public void onVarbitChange(VarbitChanged event)
+		Widget widget = event.getWidget();
+		if (WorldType.isPvpWorld(client.getWorldType())
+			&& WidgetInfo.TO_GROUP(widget.getId()) == WidgetInfo.PVP_CONTAINER.getGroupId()
+			&& WidgetInfo.TO_CHILD(widget.getId()) >= WidgetInfo.PVP_CONTAINER.getChildId())
```
is this >= really necessary? seems wonky

Nightfirecat replied:
I guess not. Checking the group is probably sufficient.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/timers/TimersPlugin.java:
```diff
@@ -146,6 +158,32 @@ public void onVarbitChange(VarbitChanged event)
+				log.debug("Left wilderness in non-PVP world, clearing Teleblock timer.");
+				removeTbTimers();
+			}
+			lastWildernessVarb = inWilderness;
```
Add empty line above

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/timers/TimersPlugin.java:
```diff
@@ -146,6 +158,32 @@ public void onVarbitChange(VarbitChanged event)
 		}
+
+		int inWilderness = client.getVar(Varbits.IN_WILDERNESS);
+		if (lastWildernessVarb != inWilderness
```
add empty line above

Nightfirecat commented:
@deathbeam Thanks for the double-check after the recent changes, addressed review comments.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/timers/TimersPlugin.java:
```diff
@@ -506,31 +542,44 @@ public void onGameTick(GameTick event)
-			removeGameTimer(DMM_FULLTB);
+			case HOPPING:
+			case LOGIN_SCREEN:
+				removeTbTimers();
```
break; after

Nightfirecat replied:
Done

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/timers/TimersPlugin.java:
```diff
@@ -506,31 +542,44 @@ public void onGameTick(GameTick event)
+			case LOGIN_SCREEN:
+				removeTbTimers();
+			case LOGGED_IN:
+				loggedInRace = true;
```
same here

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/timers/TimersPlugin.java:
```diff
@@ -506,31 +542,46 @@ public void onGameTick(GameTick event)
+				removeTbTimers();
+				break;
+			case LOGGED_IN:
+				loggedInRace = true;
```
When do you actually set this variable to false?

Nightfirecat replied:
I think this got lost during some rebasing or rework... This was meant to be set to false on game tick, so I've re-added that.
